### DISCUSSION TITLE
mamba init

### DISCUSSION THREAD

**Discussion by Jensoulou** [idaholab/moose#25703] [Q&A Getting Started]
Hello everyone, 

I am trying install MOOSE on a Mac (with Appel processor) but I got the below message when I am initializing mamba (init mamba). I tried to reinstall mamba several times and even older versions. All give me the following message: 

[mamba-init-error.txt](https://github.com/idaholab/moose/files/12858409/mamba-init-error.txt)

Can you advise me what to do?

Jens

**Comment by GiudGiud** [comment]
what does ` echo $SHELL` return ?
people have had issues with zsh and mamba

**Reply by Jensoulou** [reply]
Yesterday, it gave '-zsh'. Now, it is '/bin/zsh'.

**Reply by GiudGiud** [reply]
alright so is there a .zshenv / .zshrc file under home (cd ~/ to go there) and does it contain the mamba paragraph this init command is supposed to add

**Reply by Jensoulou** [reply]
I don't think so. I really new to this. I listed what was in the folder. And I only have other folders :-/ : 

jn@MacBook-Pro-de-Jens-2 ~ % cd ~/
jn@MacBook-Pro-de-Jens-2 ~ % ls
Applications			Music
Desktop				OneDrive - UCL
Documents			Pictures
Downloads			Public
Figures_Ink.svg			Zotero
Library				mambaforge3
MEGA				matlab_crash_dump.9718-1
Mambaforge-MacOSX-arm64.sh	miniforge3
Miniforge3-MacOSX-arm64.sh	projects
Movies

**Reply by GiudGiud** [reply]
ls wont show hidden files I think. You ll need to do `ls -a`

**Reply by Jensoulou** [reply]
There is indeed a .zshrc file: 
```
jn@MacBook-Pro-de-Jens-2 ~ % ls -a
.				.zsh_sessions
..				.zshrc
.CFUserTextEncoding		.zshrc.save
.DS_Store			Applications
.Trash				Desktop
.anaconda			Documents
.bash_profile			Downloads
.bashrc				Figures_Ink.svg
.conda				Library
.config				MEGA
.continuum			Mambaforge-MacOSX-arm64.sh
.cups				Miniforge3-MacOSX-arm64.sh
.ipynb_checkpoints		Movies
.ipython			Music
.jupyter			OneDrive - UCL
.lesshst			Pictures
.matplotlib			Public
.oracle_jre_usage		Zotero
.spyder-py3			mambaforge3
.swiftpm			matlab_crash_dump.9718-1
.tcshrc				miniforge3
.xonshrc			projects
.zsh_history
```
The file is the following: 
<img width="537" alt="Capture d’écran 2023-10-10 à 17 42 06" src="https://github.com/idaholab/moose/assets/99012294/061183d7-8bc7-4971-8df1-f641ade6d514">

**Reply by GiudGiud** [reply]
This looks fine. Ok I m out of ideas here

@milljm anything else to watch for when mamba is borked?
This isnt really a MOOSE issue btw. You should look at stackoverflow about mamba / conda

**Reply by milljm** [reply]
`/opt/anaconda3` seems out of place here.

That location will also need sudo rights, if it was not `chown`d to your user id, after its creation.

Basically, it seems like you already have a Conda installation elsewhere, that might be conflicting when trying to install this new one...

**Reply by Jensoulou** [reply]
Hey, 

Oh, I am sorry it is not the best place to ask. I appreciate your help :-)

I had Anaconda installed before trying to install Moose, but I removed it now and tried everything again. Nothing changed.

I don't find any other Conda thing elsewhere on the computer. Is there a way to be sure to remove everything Conda-related? 

As a last resort, I could reboot the computer I guess, but isn't it a bit extreme? 

Thank you so much for your help, 

Jens

**Reply by milljm** [reply]
Can you post the output that occurs when you again try to `mamba init`? The previous log indicates that Anaconda was a part of your path at the time:

```pre
<trimmed>
               CONDA_ROOT=/Users/jn/mambaforge3
           CURL_CA_BUNDLE=<not set>
               LD_PRELOAD=<not set>
                     PATH=/Users/jn/mambaforge3/bin:/opt/anaconda3/bin:/usr/local/bin:/System/Cr
                                                     ^^^^^^^^^^^^^^^ This right here is problematic
                          yptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.s
                          ecurity.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.app
                          le.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple
                          .security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Libra
                          ry/Apple/usr/bin:/Library/TeX/texbin:/usr/local/munki
       REQUESTS_CA_BUNDLE=<not set>
            SSL_CERT_FILE=<not set>
<trimmed>
```
Does the same still occur?

If so, I think its safe to delete your `~/.zshrc`. As the only contents in this file is to initialize Conda.

```bash
rm ~/.zshrc
mamba init
```

If the error persists... can you please report what `env` returns (a lot of information will be printed).

**Reply by Jensoulou** [reply]
Yes, the same occurs even though anaconda is deleted: 
[mamba_init_error_20231011.txt](https://github.com/idaholab/moose/files/12867612/mamba_init_error_20231011.txt)

So, I deleted the ~/.zshrc file: 
<img width="1073" alt="Capture d’écran 2023-10-11 à 11 34 30" src="https://github.com/idaholab/moose/assets/99012294/1016b67c-5995-481a-92f1-ddf3cc3cacaa">

Sadly, nothing is changing: 
[mamba_init_error_20231011_2.txt](https://github.com/idaholab/moose/files/12867673/mamba_init_error_20231011_2.txt)

Here is the text from 'env': 
[mamba_init_error_20231011_3.txt](https://github.com/idaholab/moose/files/12867681/mamba_init_error_20231011_3.txt)

Many thanks, 

Jens

**Reply by milljm** [reply]
Somehow, `/opt/anaconda3` still exists in your PATH. You mention:

_I had Anaconda installed before trying to install Moose, but I removed it now_ 

Do you mean to say you deleted the entire `/opt/anaconda3` directory (which is fine if you did)?

But to still have `/opt/anaconda3` in your PATH, it must be coming from somewhere. Can you post the contents of your `~/.bash_profile` and `~/.bashrc` ? I wonder if somehow ZSH is inheriting environment variables from these two source initialization files...

**Reply by Jensoulou** [reply]
I had the application Anaconda installed, I just deleted the application, like any app I don't want anymore. Should I remove other files too?

When I type 'cd /opt/anaconda3', I am going somewhere. But there are only two hidden files in it: 
<img width="296" alt="Capture d’écran 2023-10-11 à 14 12 07" src="https://github.com/idaholab/moose/assets/99012294/c72988db-203c-4f42-a87f-0d0651f767e9">

Here is the '~/.bash_profile': 
<img width="538" alt="Capture d’écran 2023-10-11 à 14 13 53" src="https://github.com/idaholab/moose/assets/99012294/56577d5f-e528-4ab6-bf94-7b2ad7f88d6a">

Here is the '~/.bashrc' (It is pretty weird...): 
<img width="367" alt="Capture d’écran 2023-10-11 à 14 15 00" src="https://github.com/idaholab/moose/assets/99012294/f6b9de39-923b-40a5-a2d4-beefb2225d62">

Many thanks :-)

**Reply by milljm** [reply]
I would delete both `~/.bash_profile` and `~/.bashrc`, since they don't appear to have any function anymore. As for why this is causing `mamba init` to crash, I am not sure... The `echo "Binjour"`... yeah that is odd! If you didn't put that there, I don't know why it would exist. Also, `echo`ing during initialization scripts _sometimes_ has the bad effect of triggering non-zero exit codes during remote operations (like ssh, scp, rsync). I would remove it personally :) however friendly it was trying to be.

Do you have any files/directories in `/usr/local/bin`?

I have one last trick to try. But even if it works, I am not sure how sustainable it would be. Try the following, and post the results. Also, its non-standard to post images of text, when a simple copy/paste will do. When we have images of text, you essentially prevent this discussion from being 'searchable' for those seeking answers to possible the same issue you are running into.


### Round 1
Close all terminal windows, and open a new one, and then:
```bash
which mamba
which python python3
env -i bash --norc --noprofile
which python python3
```
_I just want to see where python and/or `mamba` might reside before and after initializing a clean environment._


### Round 2
```bash
env -i bash --norc --noprofile
export HOME=/Users/jn
export PATH=$HOME/mambaforge3/bin:$PATH
mamba init
source activate $HOME/mambaforge3

which mamba
which python python3
```
The above will probably work (I'd hate to jinx it by saying so), but if it does not, that would mean something more complex is preventing the installation of Mamabforge3

**Reply by Jensoulou** [reply]
I deleted the ~/.bash_profile and ~/.bashrc files. 

I do have files in /usr/local/bin: 

```pre
jn@wifi-secure4-246 ~ % ls -a /usr/local/bin
.		SophosUpdate	gs		gsbj		gslj		lprsetup.sh	pf2afm		printafm	ps2pdf		ps2pdf14	ps2ps2
..		dvipdf		gs-X11		gsdj		gslp		pdf2dsc		pfbtopfa	ps2ascii	ps2pdf12	ps2pdfwr	sweep
RunSophosUpdate	eps2eps		gs-noX11	gsdj500		gsnd		pdf2ps		pphs		ps2epsi		ps2pdf13	ps2ps		unix-lpr.sh
```

ROUND 1: This is the output: 
```pre
Last login: Wed Oct 11 11:32:32 on ttys000
jn@wifi-secure4-246 ~ % which mamba
which python python3
env -i bash --norc --noprofile
which python python3
mamba not found
python not found
/usr/bin/python3

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
bash-3.2$ 
```
Round 2: Looks like it worked didn't it?: 
```pre
bash-3.2$ env -i bash --norc --noprofile

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
bash-3.2$ export HOME=/Users/jn
bash-3.2$ export PATH=$HOME/mambaforge3/bin:$PATH
bash-3.2$ mamba init

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████

no change     /Users/jn/mambaforge3/condabin/conda
no change     /Users/jn/mambaforge3/bin/conda
no change     /Users/jn/mambaforge3/bin/conda-env
no change     /Users/jn/mambaforge3/bin/activate
no change     /Users/jn/mambaforge3/bin/deactivate
no change     /Users/jn/mambaforge3/etc/profile.d/conda.sh
no change     /Users/jn/mambaforge3/etc/fish/conf.d/conda.fish
no change     /Users/jn/mambaforge3/shell/condabin/Conda.psm1
no change     /Users/jn/mambaforge3/shell/condabin/conda-hook.ps1
no change     /Users/jn/mambaforge3/lib/python3.10/site-packages/xontrib/conda.xsh
no change     /Users/jn/mambaforge3/etc/profile.d/conda.csh
modified      /Users/jn/.bash_profile

==> For changes to take effect, close and re-open your current shell. <==

Added mamba to /Users/jn/.bash_profile

==> For changes to take effect, close and re-open your current shell. <==

bash-3.2$ source activate $HOME/mambaforge3
(base) bash-3.2$ 
(base) bash-3.2$ which mamba
/Users/jn/mambaforge3/bin/mamba
(base) bash-3.2$ which python python3
/Users/jn/mambaforge3/bin/python
/Users/jn/mambaforge3/bin/python3
(base) bash-3.2$ 
```
Is everything alright now?
Thank you so much ;-)

PS: I am really sorry for the screenshots, I thought the copy/pasting would not be practical to read.

**Reply by milljm** [reply]
yeah that seems to have worked in a vanilla shell (`env -i bash --norc --noprofile`). So why/what was getting in the way before...

It says `mamba init` modified `/Users/jn/.bash_profile`. Can you paste whats in that file? I think what we will end up doing, is modifying whats there and place the contents in `~/.zshrc`

Ha, no worries! We don't like to police it, but in the end it does help others when we use text as much as possible :). Also, if you're interested, you can edit your post, ans see how I edited it, to include the fancy per-formatted text box markdown. GitHub allows for some nice fancy formatting (the tiny 'M' at the bottom right of the text box while you enter text).

**Reply by Jensoulou** [reply]
Here is the content of the `/Users/jn/.bash_profile`: 
```
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/jn/mambaforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/jn/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/jn/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/jn/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup

if [ -f "/Users/jn/mambaforge3/etc/profile.d/mamba.sh" ]; then
    . "/Users/jn/mambaforge3/etc/profile.d/mamba.sh"
fi
# <<< conda initialize <<<
```

**Reply by milljm** [reply]
Okay, that looks good. Lets copy that over to the zsh profile initialization script, and change the one thing in there that denotes it as zsh.

Hah, I just saw your text turn into proper markdown, Cheers!

Can you do the following:

```bash
cp ~/.bash_profile ~/.zshrc
```

Next, we need to edit the `~/.zshrc` to change the one instance where it says `bash` to `zsh`:

```pre
__conda_setup="$('/Users/jn/mambaforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
                                                          ^^^^ zsh
```

And finally, open a new terminal, and cross our fingers that it says `(base)` somewhere in your prompt!

**Reply by Jensoulou** [reply]
Yeaaaaah it worked :-D Thank you so much ;-)

```
Last login: Thu Oct 12 14:49:25 on ttys000
(base) jn@wifi-secure4-246 ~ % 
```

So now, I am continuing the installation of moose from that step?
```conda config --add channels https://conda.software.inl.gov/public```


Thank you very much for your precious help :-)

**Reply by milljm** [reply]
Indeed! Add the channel, and proceed with the Getting Started instructions :)

Cheers!
